backbone/src/main/java/org/researchstack/backbone/ui/task/TaskViewModel.kt

```diff
@@ -280,4 +278,47 @@ internal class TaskViewModel(val context: Application, intent: Intent) : Android
         stack.clear()
         editing = false
     }
+
+    fun checkForSaveDialog(stepResult: StepResult<*>?) {
+        if ((!isSavedDialogAppeared && !checkIfAnswersAreTheSame(stepResult) && checkIfCurrentStepIsBranchDecisionStep()) ||
+                (currentStep!!.isOptional && checkIfTheNewAnswerIsSkipWhileThePreviousIsNot(stepResult))) {
+            isSavedDialogAppeared = true
+            showSaveEditDialog.postValue(true)
+        } else {
+            nextStep()
+        }
+    }
+
+    fun saveEditDialogDismiss() {
+        isSavedDialogAppeared = false
+    }
+
+    private fun checkIfAnswersAreTheSame(currentAnswer: StepResult<*>?): Boolean {
+        return clonedTaskResultInCaseOfCancel?.let {
+            val previousAnswer = it.getStepAndResult(currentStep!!.identifier).second
+            return previousAnswer?.let { previousAnswerResult ->
+                previousAnswerResult.result == currentAnswer?.let { currentAnswerResult ->
```

Comment:
If I am not mistaken, you can replace this last `let` with:

`previousAnswerResult.result == currentAnswer?.result ?: false`